File: agent-adapter/python/turboflow_adapter/adapter.py
"""Main Adapter class — high-level API."""

from __future__ import annotations

import os

from turboflow_adapter.backend import AgentBackend
from turboflow_adapter.errors import BackendNotInstalledError, CapabilityNotSupportedError
from turboflow_adapter.logger import get_logger
from turboflow_adapter.registry import registry, create_backend
from turboflow_adapter.types import (
    BackendInfo,
    ExecOptions,
    ExecResult,
    HealthCheckResult,
    McpServer,
)

log = get_logger("tf-adapter")


class AgentAdapter:
    """High-level adapter that wraps the registry and active backend."""

    def __init__(self, backend_id: str | None = None) -> None:
        id = backend_id or os.environ.get("TURBOFLOW_AGENT_BACKEND", "claude")
        self._backend = create_backend(id)
        log.info("Initialized with backend: %s", self._backend.name)

    @property
    def active_backend(self) -> AgentBackend:
        return self._backend

    @property
    def active_backend_id(self) -> str:
        return self._backend.id

    # ── Backend management ───────────────────────────────────────────────

    async def switch_backend(self, id: str) -> None:
        new_backend = create_backend(id)
        if not await new_backend.is_installed():
            raise BackendNotInstalledError(id)
        self._backend = new_backend
        log.info("Switched to backend: %s", new_backend.name)

    def list_backends(self) -> list[BackendInfo]:
        return registry.list_all()

    async def status(self) -> list[BackendInfo]:
        results = []
        for id in registry.list_ids():
            backend = create_backend(id)
            info = backend.to_info()
            info.installed = await backend.is_installed()
            info.version = await backend.version() if info.installed else None
            info.active = id == self.active_backend_id
            results.append(info)
        return results

    # ── Agent operations ─────────────────────────────────────────────────

    async def exec(self, prompt: str, options: ExecOptions | None = None) -> ExecResult:
        return await self._backend.exec(prompt, options)

    async def version(self) -> str | None:
        return await self._backend.version()

    async def health_check(self) -> HealthCheckResult:
        return await self._backend.health_check()

    async def install(self) -> None:
        return await self._backend.install()

    # ── MCP ──────────────────────────────────────────────────────────────

    async def mcp_add(self, server: McpServer) -> None:
        if not self._backend.has_capability("mcp"):
            log.warning("Backend '%s' does not support MCP", self._backend.name)
            return
        await self._backend.mcp_add(server)

    async def mcp_remove(self, name: str) -> None:
        if not self._backend.has_capability("mcp"):
            return
        await self._backend.mcp_remove(name)

    async def mcp_list(self) -> list[McpServer]:
        if not self._backend.has_capability("mcp"):
            return []
        return await self._backend.mcp_list()

    # ── Capabilities ─────────────────────────────────────────────────────

    def has_capability(self, cap: str) -> bool:
        return self._backend.has_capability(cap)

    def require_capability(self, cap: str) -> None:
        if not self.has_capability(cap):
            raise CapabilityNotSupportedError(self.active_backend_id, cap)

    def resolve_model(self, model: str) -> str:
        return self._backend.resolve_model(model)

File: agent-adapter/ts/src/adapter.ts
// =============================================================================
// TurboFlow Agent Adapter — Main Adapter Class
//
// High-level API that wraps the registry and active backend.
// This is the primary entry point for programmatic usage.
// =============================================================================

import type {
  BackendId,
  BackendInfo,
  ExecOptions,
  ExecResult,
  HealthCheckResult,
  McpServer,
} from './types.js';
import type { AgentBackend } from './backend.js';
import { registry, createBackend } from './registry.js';
import { BackendNotInstalledError, CapabilityNotSupportedError } from './errors.js';
import { Logger } from './logger.js';

export class AgentAdapter {
  private backend: AgentBackend;
  private readonly log = new Logger('adapter');

  constructor(backendId?: BackendId | string) {
    const id = backendId ?? process.env.TURBOFLOW_AGENT_BACKEND ?? 'claude';
    this.backend = createBackend(id);
    this.log.info(`Initialized with backend: ${this.backend.name}`);
  }

  /** Get the active backend */
  get activeBackend(): AgentBackend {
    return this.backend;
  }

  /** Get the active backend ID */
  get activeBackendId(): string {
    return this.backend.toInfo().id;
  }

  // ── Backend management ──────────────────────────────────────────────────

  /** Switch to a different backend */
  async switchBackend(id: BackendId | string): Promise<void> {
    const newBackend = createBackend(id);
    const installed = await newBackend.isInstalled();

    if (!installed) {
      throw new BackendNotInstalledError(id);
    }

    this.backend = newBackend;
    this.log.info(`Switched to backend: ${newBackend.name}`);
  }

  /** List all available backends */
  listBackends(): BackendInfo[] {
    return registry.listAll();
  }

  /** Get status of all backends (installed, version, etc.) */
  async status(): Promise<Array<BackendInfo & { installed: boolean; version: string | null }>> {
    const results = [];
    for (const id of registry.listIds()) {
      const backend = createBackend(id);
      const installed = await backend.isInstalled();
      const version = installed ? await backend.version() : null;
      results.push({
        ...backend.toInfo(),
        installed,
        version,
        active: id === this.activeBackendId,
      });
    }
    return results;
  }

  // ── Agent operations ────────────────────────────────────────────────────

  /** Execute a prompt through the active agent */
  async exec(prompt: string, options?: ExecOptions): Promise<ExecResult> {
    this.log.info(`Executing prompt via ${this.backend.name}`, {
      model: options?.model,
      headless: options?.headless,
      files: options?.files?.length,
    });

    const startTime = Date.now();
    const result = await this.backend.exec(prompt, options);

    this.log.info(`Execution complete`, {
      exitCode: result.exitCode,
      durationMs: result.durationMs,
      timedOut: result.timedOut,
    });

    return result;
  }

  /** Get agent version */
  async version(): Promise<string | null> {
    return this.backend.version();
  }

  /** Health check */
  async healthCheck(): Promise<HealthCheckResult> {
    return this.backend.healthCheck();
  }

  /** Install the active backend */
  async install(): Promise<void> {
    return this.backend.install();
  }

  // ── MCP operations ──────────────────────────────────────────────────────

  /** Register an MCP server */
  async mcpAdd(server: McpServer): Promise<void> {
    if (!this.backend.hasCapability('mcp')) {
      this.log.warn(`Backend '${this.backend.name}' does not support MCP — skipping`);
      return;
    }
    return this.backend.mcpAdd(server);
  }

  /** Remove an MCP server */
  async mcpRemove(name: string): Promise<void> {
    if (!this.backend.hasCapability('mcp')) return;
    return this.backend.mcpRemove(name);
  }

  /** List MCP servers */
  async mcpList(): Promise<McpServer[]> {
    if (!this.backend.hasCapability('mcp')) return [];
    return this.backend.mcpList();
  }

  // ── Capability checks ──────────────────────────────────────────────────

  /** Check if the active backend supports a capability */
  hasCapability(cap: string): boolean {
    return this.backend.hasCapability(cap as any);
  }

  /** Require a capability — throws if not supported */
  requireCapability(cap: string): void {
    if (!this.hasCapability(cap)) {
      throw new CapabilityNotSupportedError(this.activeBackendId, cap);
    }
  }

  // ── Model resolution ───────────────────────────────────────────────────

  /** Resolve a model tier name to a backend-specific model string */
  resolveModel(model: string): string {
    return this.backend.resolveModel(model);
  }
}
